js/ui/marker.js

```diff
@@ -75,19 +78,91 @@ Marker.prototype = {
      * @param {LngLat} lnglat
      * @returns {Marker} `this`
      */
-    setLngLat: function(lnglat) {
+    setLngLat: function (lnglat) {
         this._lngLat = LngLat.convert(lnglat);
+        if (this._popup) this._popup.setLngLat(this._lngLat);
         this._update();
         return this;
     },
 
-    getElement: function() {
+    getElement: function () {
         return this._el;
     },
 
-    _update: function() {
+    /**
+     * Binds a Popup to the Marker
+     * @param {Popup=} popup an instance of the `Popup` class. If undefined or null, any popup 
+     * set on this `Marker` instance is unset
+     * @returns {Marker} `this`
+     */
+
+    setPopup: function (popup) {
+        if (popup == null) {
+            this._closePopup();
+            delete this._popupHandlersAdded;
```

Comment:
Instead of tracking state in `_popupHandlersAdded`, call`removeEventListener` when the popup is removed. This will prevent memory leaks and unexpected behavior if the `Marker` continues to be used without a popup. You'll need to save the `this`-bound version of `_openPopup`, say by using `util.bindAll(['_openPopup'], this)`, so that it can be passed to `removeEventListener`.